drawImage: clip source image when source rectangle outside of source image.

brentfulgham opened the pull request:
#### a31e58979ddb60c1b9332d9ce5a7b52bdc34c1d7
<pre>
drawImage: clip source image when source rectangle outside of source image.
<a href="https://bugs.webkit.org/show_bug.cgi?id=116991">https://bugs.webkit.org/show_bug.cgi?id=116991</a>
&lt;rdar://problem/97222886&gt;

Reviewed by Darin Adler.

Based on original work by Arnaud Renevier &lt;a.renevier@sisa.samsung.com&gt;

Add a static function normalizeSourceAndDestination to normalize source and
destination rectangles, depending on image/canvas/video rectangle. That method
is used to make dimension checking between different drawImage methods consistent.

If image rectangle does not contain source rectangle, intersect source
rectangle with image rectangle, and shrink destination size by the
same ratio source rectangle was shrinked.

Test: fast/canvas/drawImage-source-rect-outside-image.html

* LayoutTests/fast/canvas/drawImage-source-rect-outside-image-expected.html: Added.
* LayoutTests/fast/canvas/drawImage-source-rect-outside-image.html: Added.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::normalizeSourceAndDestination):
(WebCore::CanvasRenderingContext2DBase::drawImage):

Canonical link: <a href="https://commits.webkit.org/252731@main">https://commits.webkit.org/252731@main</a>
</pre>

brentfulgham commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -1643,9 +1662,17 @@ ExceptionOr<void> CanvasRenderingContext2DBase::drawImage(HTMLVideoElement& vide
+    if (normalizedDstRect.isEmpty())
         return { };
 
+
```
Remove this whitespace.

darinadler commented on LayoutTests/fast/canvas/drawImage-source-rect-outside-image.html:
```diff
@@ -0,0 +1,25 @@
+
+    srcCtx.arc(width / 2, height / 2, Math.min(width / 2, height / 2), 0, Math.PI * 2)
+    srcCtx.fill();
+    ctx.drawImage(srcCanvas, width / 2, height / 2, width, height, 0, 0, width, height);
```
I’m really surprised there is no coverage for this in WPT. Should we put a test in there?

brentfulgham replied:
Yes, I think so. I'll work with Tim to get a "real" WPT case together.

pascoej commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -1659,24 +1686,24 @@ ExceptionOr<void> CanvasRenderingContext2DBase::drawImage(HTMLVideoElement& vide
 #if USE(CG)
```
Informal:

`    bool repaintEntireCanvas = rectContainsCanvas(dstRect);`
Does this need normalizedDstRect?

brentfulgham replied:
Oh! I'll bet it should! We probably need another test case to support it.

brentfulgham committed:
drawImage: clip source image when source rectangle outside of source image.
https://bugs.webkit.org/show_bug.cgi?id=116991
<rdar://problem/97222886>

Reviewed by Darin Adler.

Based on original work by Arnaud Renevier <a.renevier@sisa.samsung.com>

Add a static function normalizeSourceAndDestination to normalize source and
destination rectangles, depending on image/canvas/video rectangle. That method
is used to make dimension checking between different drawImage methods consistent.

If image rectangle does not contain source rectangle, intersect source
rectangle with image rectangle, and shrink destination size by the
same ratio source rectangle was shrinked.

Test: fast/canvas/drawImage-source-rect-outside-image.html

* LayoutTests/fast/canvas/drawImage-source-rect-outside-image-expected.html: Added.
* LayoutTests/fast/canvas/drawImage-source-rect-outside-image.html: Added.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::normalizeSourceAndDestination):
(WebCore::CanvasRenderingContext2DBase::drawImage):

Canonical link: https://commits.webkit.org/252731@main
